// ngx-filesaver/esm2015/src/filesaver.directive.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { Directive, ElementRef, Input, HostListener, EventEmitter, Output } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { FileSaverService } from './filesaver.service';
let FileSaverDirective = class FileSaverDirective {
    constructor(el, fss, httpClient) {
        this.el = el;
        this.fss = fss;
        this.httpClient = httpClient;
        this.method = 'GET';
        this.success = new EventEmitter();
        this.error = new EventEmitter();
        if (!fss.isFileSaverSupported) {
            el.nativeElement.classList.add(`filesaver__not-support`);
        }
    }
    getName(res) {
        return decodeURI(this.fileName || res.headers.get('filename') || res.headers.get('x-filename'));
    }
    _click() {
        if (!this.fss.isFileSaverSupported) {
            return;
        }
        let req = this.http;
        if (!req) {
            const params = new HttpParams();
            const query = this.query || {};
            // tslint:disable-next-line:forin
            for (const item in query) {
                params.set(item, query[item]);
            }
            req = this.httpClient.request(this.method, this.url, {
                observe: 'response',
                responseType: 'blob',
                headers: this.header,
                params,
            });
        }
        this.setDisabled(true);
        req.subscribe(res => {
            if (res.status !== 200 || res.body.size <= 0) {
                this.error.emit(res);
                return;
            }
            this.fss.save(res.body, this.getName(res));
            this.success.emit(res);
        }, err => this.error.emit(err), () => this.setDisabled(false));
    }
    setDisabled(status) {
        const el = this.el.nativeElement;
        el.disabled = status;
        el.classList[status ? 'add' : 'remove'](`filesaver__disabled`);
    }
};
__decorate([
    Input(),
    __metadata("design:type", Object)
], FileSaverDirective.prototype, "method", void 0);
__decorate([
    Input(),
    __metadata("design:type", Observable)
], FileSaverDirective.prototype, "http", void 0);
__decorate([
    Input(),
    __metadata("design:type", Object)
], FileSaverDirective.prototype, "query", void 0);
__decorate([
    Input(),
    __metadata("design:type", Object)
], FileSaverDirective.prototype, "header", void 0);
__decorate([
    Input(),
    __metadata("design:type", String)
], FileSaverDirective.prototype, "url", void 0);
__decorate([
    Input(),
    __metadata("design:type", String)
], FileSaverDirective.prototype, "fileName", void 0);
__decorate([
    Output(),
    __metadata("design:type", Object)
], FileSaverDirective.prototype, "success", void 0);
__decorate([
    Output(),
    __metadata("design:type", Object)
], FileSaverDirective.prototype, "error", void 0);
__decorate([
    HostListener('click'),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", void 0)
], FileSaverDirective.prototype, "_click", null);
FileSaverDirective = __decorate([
    Directive({
        selector: '[fileSaver]',
        exportAs: 'fileSaver',
    }),
    __metadata("design:paramtypes", [ElementRef, FileSaverService, HttpClient])
], FileSaverDirective);
export { FileSaverDirective };
