// ngx-audio-player/esm5/lib/component/base/base-audio-player-components.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ViewChild, ElementRef } from "@angular/core";
var BaseAudioPlayerFunctions = /** @class */ (function () {
    function BaseAudioPlayerFunctions() {
        this.loaderDisplay = false;
        this.isPlaying = false;
        this.currentTime = 0;
        this.volume = 0.1;
        this.duration = 0.01;
    }
    /**
     * @param {?} event
     * @return {?}
     */
    BaseAudioPlayerFunctions.prototype.currTimePosChanged = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.player.nativeElement.currentTime = event.value;
    };
    /**
     * @return {?}
     */
    BaseAudioPlayerFunctions.prototype.bindPlayerEvent = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.player.nativeElement.addEventListener('playing', (/**
         * @return {?}
         */
        function () {
            _this.isPlaying = true;
            _this.duration = Math.floor(_this.player.nativeElement.duration);
        }));
        this.player.nativeElement.addEventListener('pause', (/**
         * @return {?}
         */
        function () {
            _this.isPlaying = false;
        }));
        this.player.nativeElement.addEventListener('timeupdate', (/**
         * @return {?}
         */
        function () {
            _this.currentTime = Math.floor(_this.player.nativeElement.currentTime);
        }));
        this.player.nativeElement.addEventListener('volume', (/**
         * @return {?}
         */
        function () {
            _this.volume = Math.floor(_this.player.nativeElement.volume);
        }));
        this.player.nativeElement.addEventListener('loadstart', (/**
         * @return {?}
         */
        function () {
            _this.loaderDisplay = true;
        }));
        this.player.nativeElement.addEventListener('loadeddata', (/**
         * @return {?}
         */
        function () {
            _this.loaderDisplay = false;
            _this.duration = Math.floor(_this.player.nativeElement.duration);
        }));
    };
    ;
    /**
     * @return {?}
     */
    BaseAudioPlayerFunctions.prototype.playBtnHandler = /**
     * @return {?}
     */
    function () {
        if (this.loaderDisplay) {
            return;
        }
        if (this.player.nativeElement.paused) {
            this.player.nativeElement.play(this.currentTime);
        }
        else {
            this.currentTime = this.player.nativeElement.currentTime;
            this.player.nativeElement.pause();
        }
    };
    ;
    /**
     * @return {?}
     */
    BaseAudioPlayerFunctions.prototype.play = /**
     * @return {?}
     */
    function () {
        var _this = this;
        setTimeout((/**
         * @return {?}
         */
        function () {
            _this.player.nativeElement.play();
        }), 0);
    };
    ;
    /**
     * @return {?}
     */
    BaseAudioPlayerFunctions.prototype.toggleVolume = /**
     * @return {?}
     */
    function () {
        if (this.volume === 0) {
            this.setVolume(1.0);
        }
        else {
            this.setVolume(0);
        }
    };
    /**
     * @private
     * @param {?} vol
     * @return {?}
     */
    BaseAudioPlayerFunctions.prototype.setVolume = /**
     * @private
     * @param {?} vol
     * @return {?}
     */
    function (vol) {
        this.volume = vol;
        this.player.nativeElement.volume = this.volume;
    };
    BaseAudioPlayerFunctions.propDecorators = {
        player: [{ type: ViewChild, args: ['audioPlayer', { static: true },] }]
    };
    return BaseAudioPlayerFunctions;
}());
export { BaseAudioPlayerFunctions };
if (false) {
    /** @type {?} */
    BaseAudioPlayerFunctions.prototype.player;
    /** @type {?} */
    BaseAudioPlayerFunctions.prototype.timeLineDuration;
    /** @type {?} */
    BaseAudioPlayerFunctions.prototype.loaderDisplay;
    /** @type {?} */
    BaseAudioPlayerFunctions.prototype.isPlaying;
    /** @type {?} */
    BaseAudioPlayerFunctions.prototype.currentTime;
    /** @type {?} */
    BaseAudioPlayerFunctions.prototype.volume;
    /** @type {?} */
    BaseAudioPlayerFunctions.prototype.duration;
    /* Skipping unhandled member: ;*/
    /* Skipping unhandled member: ;*/
    /* Skipping unhandled member: ;*/
}
